gdpr/models.py

```diff
@@ -21,27 +33,106 @@ def create_from_purpose_slug(self, purpose_slug, source_object, issued_at=None,
                 expires_at=issued_at + purpose.expiration_timedelta,
                 tag=tag,
                 purpose_slug=purpose_slug,
-                source_object=source_object
+                source_object=source_object,
+                is_active=True
             )
             for related_object in related_objects or ():
                 legal_reason.related_objects.create(object=related_object)
             return legal_reason
         except KeyError:
             raise KeyError('Purpose with slug {} does not exits'.format(purpose_slug))
 
+    def create_or_update_from_purpose_slug(self, purpose_slug, source_object, issued_at=None, tag=None,
+                                           related_objects=None):
+        """
+        Update (or create, if it doesn't exist) a LegalReason with purpose slug for concrete object instance
+
+        @param purpose_slug: String of Legal Reason purpose
+        @param source_object: Source object this Legal Reason is related to
+        @param issued_at: When the Legal Reason consent was given
+        @param tag:
+        @param related_objects:
+
+        @return: Legal Reason
+        """
+        if self.filter_source_instance(source_object).filter(purpose_slug=purpose_slug).exists():
+            return self.refresh_for_source_by_slug(purpose_slug, source_object, tag, related_objects)
+        else:
+            return self.create_from_purpose_slug(purpose_slug, source_object, issued_at, tag, related_objects)
+
+    def refresh_for_source_by_slug(self, purpose_slug, source_object, tag=None, related_objects=None):
+        """
+        Update Legal Reason identified by purpose slug for source object
+
+        @param purpose_slug: String of Legal Reason purpose
+        @param source_object: Source object this Legal Reason is related to
+        @param tag:
+        @param related_objects:
+
+        @return: Legal Reason
+        """
+        assert self.filter_source_instance(source_object).filter(purpose_slug=purpose_slug).exists()
+        purpose = purposes_map[purpose_slug]
+
+        self.filter_source_instance(source_object).filter(purpose_slug=purpose_slug).update(
+            expires_at=now() + purpose.expiration_timedelta,
+            tag=tag,
+            is_active=True
+        )
+        for related_object in related_objects or ():
+            legal_reason.related_objects.create(object=related_object)
+
+        try:
+            return self.filter_source_instance(source_object).filter(purpose_slug=purpose_slug).get()
```

Comment:
Proč? nemáš to ošetřené tím unikátním klíčem?

Reply:
Navíc nahoře máš prekticky to samo v assertu

Reply:
OK, tohle tedy bude vypadat jinak.